### PR TITLE
Fix AnalysisMode propagation in NamedAnalyzer

### LINKED ISSUE #44625: Updateable analyzers are not rejected when used as index time analyzers
Analyzers that are only allowed to run at search time (e.g. because they are marked as "updateable") should not be allowed as regular analyzers on a field. E.g. the following should throw an error:
```
PUT /test
{
  "settings": {
    "index": {
      "analysis": {
        "analyzer": {
          "synonym": {
            "tokenizer": "whitespace",
            "filter": ["synonym"]
          }
        },
        "filter": {
          "synonym": {
            "type": "synonym",
            "synonyms_path": "synonyms.txt",
            "updateable": true
          }
        }
      }
    }
  }
}

POST /synonym_test/_mapping
{
  "properties": {
    "text_field" : {
      "type": "text",
      "analyzer": "synonym"
    }
  }
}
```
But currently doesn't. The reason seems to be that we don't propagate the search mode correctly when wrapping it in a NamedAnalyzer. I'm working on a fix.

### PR DESCRIPTION
**Pull request by cbuescher**
NamedAnalyzer should return the same AnalysisMode than any custom analyzer it
wraps, otherwise AnalysisMode.ALL. This used to be only CustomAnalyzer in the
past, but with the introduction of the ReloadableCustomAnalyzer this needs to be
added as an option where the analysis mode gets propagated.

Closes #44625

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/analysis/NamedAnalyzer.java [hunk 1]
```diff
@@ -55,6 +55,8 @@ public NamedAnalyzer(String name, AnalyzerScope scope, Analyzer analyzer) {
         this.positionIncrementGap = positionIncrementGap;
         if (analyzer instanceof org.elasticsearch.index.analysis.CustomAnalyzer) {
             this.analysisMode = ((org.elasticsearch.index.analysis.CustomAnalyzer) analyzer).getAnalysisMode();
+        } else if (analyzer instanceof org.elasticsearch.index.analysis.ReloadableCustomAnalyzer) {
+            this.analysisMode = ((org.elasticsearch.index.analysis.ReloadableCustomAnalyzer) analyzer).getAnalysisMode();
         } else {
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/analysis/NamedAnalyzer.java]: Maybe instead do an instanceof check for AnalyzerComponentsProvider and then call `getComponents().getAnalysisMode()`?

**Review by romseygeek — CHANGES_REQUESTED** [review]
Good catch, I have one suggested change that I think will be neater than lists of instanceof checks

**Comment by cbuescher** [comment]
@romseygeek thanks for the review, I pushed an update addressing your comment.

File: server/src/main/java/org/elasticsearch/index/analysis/NamedAnalyzer.java (hunk 1 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/analysis/NamedAnalyzer.java]: Oh for Java pattern matching... thanks!

**Review by romseygeek — APPROVED** [review]

**Pull request merged**